### ISSUE TITLE
 alert prow plugin for no cla PR merges

### ISSUE THREAD

**Issue by pacoxu** [kubernetes-sigs/prow#279]
https://kubernetes.slack.com/archives/C01672LSZL0/p1726321799185879

A prow plugin that can support sending alert to slack when a PR with `cncf-cla: no` label was merged.

- like https://github.com/search?q=org%3Akubernetes+is%3Apr+is%3Amerged+label%3A%22cncf-cla%3A+no%22+created%3A%3E2023-01-01+&type=pullrequests

**Comment by petr-muller** [comment]
In this a scope-limited form, this should not be hard to write. Get a GH webhook call about a merge, inspect the PR, send a notification (we already have [Slack client code](https://github.com/kubernetes-sigs/prow/tree/main/pkg/slack) in Prow).

Notification systems love to grow in scope, though :thinking: Some people will almost certainly want notifications about other events. Some people will almost certainly want notifications to different destinations than Slack. Would we need per-org / per-repo configuration surface? Like disabling the notifications for certain repos, or routing the notifications to different slacks based on repos?

Given the current state of Prow maintenance, I think we should prefer building a very simple, single-purpose thing rather than trying to come up with a new generic notification system.

One alternative could be something that simply publishes interesting metrics through Prometheus and then we could set up standard alertmanager-backed alerts for events like this.

**Comment by pacoxu** [comment]
> One alternative could be something that simply publishes interesting metrics through Prometheus and then we could set up standard alertmanager-backed alerts for events like this.

This seems to be a better solution to me, as the `cncf-cla： no` is not a general case and only occurs several time a year. 
- Adding a github action like thing for all PRs seems to be too heavy
- An alert may be a lightweight solution.

If we want to go this direction, is there an example to add an alert policy?

**Comment by pacoxu** [comment]
@dims @BenTheElder which solution do you prefer?

**Comment by dims** [comment]
@pacoxu see example in https://kubernetes.slack.com/archives/CJH2GBF7Y/p1724136124406689

<img width="794" alt="image" src="https://github.com/user-attachments/assets/25314eba-1097-4a4c-af01-8a4fe9998d0f">

An alert like that looks like a great thing to have

**Comment by BenTheElder** [comment]
We shouldn't rely on the label, the status is the source of truth.

Statuses can change _after_ PRs merge, so the alert is only a best effort warning in any case.

If we want to be 100% sure, the CNCF CLA tooling could run an audit mode and scan all commits in the repo against their CLA database (which would surely be cheaper than us attempting to scan the github API for commit statuses).

For a weak check:
> Given the current state of Prow maintenance, I think we should prefer building a very simple, single-purpose thing rather than trying to come up with a new generic notification system.

Yes, I think we could reuse some tooling from the slack alerts but I think it should be a single purpose plugin if we implement this, to keep things as simple as possible.

I think the problem with using a metric is then we still need to track down _where_ the bad merge happened, and now we have large cardinality on the metric if we're recording it there. Versus if we just push out a slack alert (and/or completely offload this to CNCF auditing).

**Comment by BenTheElder** [comment]
(Poking some CNCF folks about the subject ...)

**Comment by pacoxu** [comment]
link to https://github.com/kubernetes/community/issues/7800

**Comment by BenTheElder** [comment]
I disagree strongly with the linked issue and I think that's an unrelated topic for now.

**Comment by petr-muller** [comment]
/remove-lifecycle rotten

This would fit into an existing [slackevents](https://github.com/kubernetes-sigs/prow/tree/main/pkg/plugins/slackevents) plugin (which is actually what backs the force push alert linked [above](https://github.com/kubernetes-sigs/prow/issues/279#issuecomment-2368159969)).

It is somewhat straightforward for repositories that use the `merge` method to merge PRs - merged commits will inherit the status context placed on the PR, so it is probably enough to exempt bot-authored merge commits from checking the  green CLA status presence and it should work fine.

It is harder for repositories where the status context on the original commit is lost after merge because the repo uses a `rebase` or `squash` merge strategy. It could be possible to identify the original PR and get the original status context on the pre-merge commit but it gets a bit hairy here.

**Comment by BenTheElder** [comment]
/lifecycle frozen
/help
xref: https://github.com/kubernetes/community/pull/8447

**Comment by AaruniAggarwal** [comment]
Hi @petr-muller, I’m interested in picking this up. Is this issue open for contributors to work on?
Thanks

**Comment by petr-muller** [comment]
@AaruniAggarwal yep, feel free to give it a shot. https://github.com/kubernetes-sigs/prow/issues/279#issuecomment-2646009816 gives some guidance. Extending the `slackevents` plugin seems to be a reasonable direction.

Seems like `PullRequestEvent` webhook payload will carry the original PR SHA which could be used to obtain the original status placed on the commit,.

/area plugins
